### DISCUSSION TITLE
How to take NA values into account in mean aggregation?

### DISCUSSION THREAD

**Discussion by tht-activeviam** [atoti/atoti#594] [Q&A] [answered]
Let say I created average/mean measure for distinct TRADE_ID count at DEAL_MONTH scope.
```
tt.agg.mean(m['TradeId.Distinct Count'], scope=level['DEAL_MONTH'])
```
And assume that m['TradeId.Distinct Count'] of 'April 2022' is NA. The mean aggregation ignore the 'April 2022'  in that case. 
What if we wanted to include 'April 2022' in final mean aggregation?
If the trade count is zero for that month, the result of mean aggregation will be different.
Can we set default value for measure? eg. `agg.distinct_count(table['TRADE_ID'], default=0)`

By the way, Excel function `AVERAGE` consider cells with zero value.
To ignore the zero value cell, you have to use `AVERAGEIF(range,"<>0")`.

**Comment by tibdex** [comment]
> assume that m['TradeId.Distinct Count'] of 'April 2022' is NA

`"N/A"`, `None` or `NaN`?

If it's `None` you can try:

```diff
- tt.agg.mean(m['TradeId.Distinct Count'], scope=level['DEAL_MONTH'])
+ tt.agg.mean(tt.where(m['TradeId.Distinct Count'] != None, m['TradeId.Distinct Count'], 0.0), scope=level['DEAL_MONTH'])
```

If it doesn't fix your issue can you share a minimal reproducible example?

**Comment by OPeyrusse** [comment]
It is common in SQL database systems performing aggregations to ignore NULL values
See this documentation from PostgreSQL https://www.postgresql.org/docs/14/functions-aggregate.html
>sum ( smallint ) → bigint
sum ( integer ) → bigint
sum ( bigint ) → numeric
Computes the sum of the non-null input values.

In that respect, atoti is compliant. The best solution is to manually deal with those values before performing the aggregation.

**Comment by tht-activeviam** [comment]
Hi,
I agree that `mean` API ignore the empty(null) values which is the same in Excel as well.
However, I still couldn't replace those empty values with zero. 
```python
# Atoti version: 0.6.4

m['Trade.COUNT'] = tt.agg.count_distinct(data['TRADE_ID'])
m['Client.COUNT'] = tt.agg.count_distinct(data['SCI_LEID'])

# I used > 0 to avoid Null/None/Empty confusion
m['Trade.YTD AVG'] = tt.agg.mean(tt.where(m['Trade.COUNT'] > 0, m['Trade.COUNT'], 0), scope=tt.scope.origin(lvl['DEAL_MONTH_YEAR', 'DEAL_MONTH_YEAR']))
m['Client.YTD AVG'] = tt.agg.mean(tt.where(m['Client.COUNT'] > 0, m['Client.COUNT'], 0), scope=tt.scope.origin(lvl['DEAL_MONTH_YEAR', 'DEAL_MONTH_YEAR']))
```
See below the row `REPO` and column `INTERDEALER -> APR 2022`. There is no trade in that, so the `Trade.COUNT` is empty. But I need to display that as zero for `Trade.YTD AVG` so that the total AVG calculation will include all months.
![image](https://user-images.githubusercontent.com/41277323/170437814-7e1bf759-3e74-4601-b48c-d9973060b8fe.png)

**Comment by tht-activeviam** [comment, marked as answer]
I've created the notebook to reproduce. => https://github.com/tht-activeviam/perso-tht-atoti-playground/blob/main/tt.agg.mean%20Demo.ipynb

`Atoti version: 0.6.5` (shouldn't be matter anyway)

**Reply by tibdex** [reply]
Looks like your repo is private.

**Reply by tht-activeviam** [reply]
> Looks like your repo is private.

please try again.

**Reply by tibdex** [reply]
Your data doesn't have trades for `INTERDEALER/FEB 2022` so you can't make an aggregate appear for this combination with the code you shared. One way to achieve what I think you want would be to do something like this:

```python
import atoti as tt
import pandas as pd

data_df = pd.DataFrame({
    "Trade Id": ["T001", "T002", "T003"], 
    "Contract Type": ["FRA", "IRS", "ETD"],
    "Marketer Name": ["CLIENT_BUSINESS", "INTERDEALER", "INTERDEALER"],
    "Deal Month": ["JAN 2022", "FEB 2022", "MAR 2022"]
})
data_df

session = tt.Session()
table = session.read_pandas(data_df, keys=["Trade Id"], table_name="Data")
cube = session.create_cube(table, 'DataCube')
m, h, lvl = cube.measures, cube.hierarchies, cube.levels

# Could also be replaced by `Cube.create_parameter_hierarchy_from_members()` with an exhaustive list of all the months to ensure that it's really dense regardless of the months in the facts.
cube.create_parameter_hierarchy_from_column("Dense Date", column=table["Deal Month"])
h["Dense Date"].slicing = False

h["Deals"] = [lvl["Marketer Name"], lvl["Deal Month"]]
del h["Deal Month"]

m["Trade.COUNT"] = tt.agg.count_distinct(table["Trade Id"])

m["Dense Trade.COUNT"] = tt.at(m["Trade.COUNT"], {lvl["Deal Month"]: lvl["Dense Date"]})
m["Non-nullable Dense Trade.COUNT"] = tt.where(m["Dense Trade.COUNT"] != None, m["Dense Trade.COUNT"], 0)
m["Trade.AVG"] = tt.agg.mean(m["Non-nullable Dense Trade.COUNT"], scope=tt.scope.origin(lvl["Deal Month"], lvl["Dense Date"]))

cube.query(m["Trade.AVG"], levels=[lvl["Contract Type"], lvl["Deal Month"]])
```

**Reply by tht-activeviam** [reply]
Thanks Thibault. I will try it out.

**Reply by tht-activeviam** [reply]
Hi @tibdex, 
I still have the same problem using `parameter_hierarchy_from_column`(or even `parameter_hierarchy_from_members`)
Notebook is updated here => https://github.com/tht-activeviam/perso-tht-atoti-playground/blob/main/tt.agg.mean%20Demo.ipynb

**Reply by tibdex** [reply]
In the code I adapted from your notebook, the result of the query is:

```
                                         Trade.AVG
Contract Type Marketer Name   Deal Month          
ETD           INTERDEALER     MAR 2022         .33
FRA           CLIENT_BUSINESS JAN 2022         .33
IRS           INTERDEALER     FEB 2022         .33
```

Since you have only 1 trade per **Contract Type** / **Marketer Name** and 3 months of data the **Trade.AVG** is 1/3 everywhere. Isn't this what you expect?

**Reply by tht-activeviam** [reply]
Thanks @tibdex . That looks correct.
